src/encoder.rs

```diff
@@ -0,0 +1,248 @@
+// Copyright 2016 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::io::Write;
+
+use errors::{Result, Error};
+use proto::MetricFamily;
+use proto::{self, MetricType};
+
+pub trait Encoder {
+    /// `encode` converts a slice of MetricFamily proto messages into target
+    /// format and writes the resulting lines to `writer`. It returns the number
+    /// of bytes written and any error encountered. This function does not
+    /// perform checks on the content of the metric and label names,
+    /// i.e. invalid metric or label names will result in invalid text format
+    /// output.
+    fn encode(&self, &[MetricFamily], &mut Write) -> Result<usize>;
```

Comment:
No, it returns the number of written bytes. Why? Because `Content-Length` in HTTP.

It is difficult for out caller to compute precise number of written bytes.